### PR TITLE
Smarter buildenv

### PR DESCRIPTION
**Pull request by stormbeta**
This enables experimental support for dependent values using the buildEnv extension.

A common problem I've seen is that a project will need to alter the branch or release detection pattern, but this doesn't always update dependent values correctly leading to confusing behavior.

Moreover, having to allow those things to be overridden directly means we have to delay modifying the project version in the versioning plugins until afterEvaluate, which is a problem as many projects reference the version string during the configuration phase.

To solve this, I've added experimental support for value dependencies using buildEnv.

Basically, it's now possible to attach listeners to changes on buildEnv objects. In addition, you can "automatically" bind buildEnv properties to other buildEnv properties, e.g. the `release` property is set to automatically depend on changes to `branch`.

### CODE REVIEW

**Comment by stormbeta** [comment]
@dougborg, @sgoings 

Example using the buildnumber plugin:

``` groovy

version = '0.5.0'           //version == '0.5.0-DEV'
buildEnv.ci = true          //version == '0.5.0-LOCAL'
buildEnv.buildNumber = '99' //version == '0.5.0-99'

```

Example using the snapshots plugin:

``` groovy

version = '0.5.0'          //version == '0.5.0-SNAPSHOT'
buildEnv.release = true    //version == '0.5.0'
buildEnv.release = false   //version == '0.5.0-SNAPSHOT'
buildEnv.branch = 'v0.5.0' //release == true && version == '0.5.0'

```

**Comment by stormbeta** [comment]
Tests are failing because I assumed the project version would have a value before the plugins get applied. This is technically fixable to be order independent, but would significantly increase the scope to handle read dependencies on non-extension values (which probably requires intercepting the property set on the gradle project object itself).
